### ISSUE TITLE
jax.random.randint cannot generate values equal to DTYPE_MAX

### ISSUE THREAD

**Issue by altche** [google/jax#5785]
According to randint docstring: `Sample uniform random values in [minval, maxval) with given shape/dtype.`

However because of the conversion here https://github.com/google/jax/blob/5bbb449ae5849c508194c8eb5b10c101f1fa22ae/jax/_src/random.py#L439 passing `maxval={dtype max value + 1}` will generate an all-zero array.
The behavior in numpy is different (and correct):
```
jax.random.randint(jax.random.PRNGKey(0), shape=(10,), minval=0, maxval=256, dtype=jnp.uint8)
>> DeviceArray([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint8)

np.random.randint(size=(10,), low=0, high=256, dtype=np.uint8)
>> array([ 87, 136, 195, 254,  95,  94, 113,   7,  66,  13], dtype=uint8)
```

A workaround is to not pass the dtype to randint (or use a dtype with a higher maximum value), and cast the output:
```
jax.random.randint(jax.random.PRNGKey(0), shape=(10,), minval=0, maxval=256).astype(jnp.uint8)
>> DeviceArray([178, 233,  99, 238,  39, 149,  46, 139,  68, 142], dtype=uint8)
```

**Comment by jakevdp** [comment]
Thinking about this a bit... this is really tricky for JAX to handle in a general way. Consider the case of `max = 1 << 64`, which is one greater than the maximum uint64. Numpy handles it as expected:
```python
>>> np.random.randint(0, 1<<64, dtype='uint64')
17392714807819977315
```
But if you increase this bound by one, numpy returns an error:
```
>>> np.random.randint(0, 1<<64 + 1, dtype='uint64')                                                                                                                         
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-30-e02298560197> in <module>
----> 1 np.random.randint(0, 1<<64 + 1, dtype='uint64')

mtrand.pyx in numpy.random.mtrand.RandomState.randint()

_bounded_integers.pyx in numpy.random._bounded_integers._rand_uint64()

ValueError: high is out of bounds for uint64
```
Neither of these numbers (`1 << 64` and `1 << 64 + 1`) are representable as 64-bit integers or floats; numpy handles these by making use of Python's built-in arbitrary precision integers, as you can see here: https://github.com/numpy/numpy/blob/5cae51e794d69dd553104099305e9f92db237c53/numpy/random/_bounded_integers.pyx.in#L190-L192

JAX cannot do this, because it compiles to XLA which only has fixed-precision integers, and has no native way to represent larger numbers such as `1 << 64`.

How might we work around this? Some possibilities:

1. Follow NumPy and do our arithmetic in Python. The downside of this is that it would only work for concrete values of minval and maxval; i.e. you could no longer jit-compile code that passes minval and maxval as JAX arrays. You might imagine getting around this by only using Python ints in non-traced code; the problem there is that this would cause `randint` to act differently depending on whether it is wrapped in JIT, which is something we like to avoid.

2. Use some sort of extended representation (i.e. a pair of int64s) to represent the minval and maxval in order to handle larger bounds. However, this would require deeper changes to JAX's JIT dispatch mechanism, because currently a Python integer passed to a jitted function is **always** converted to a dtyped integer, because it's the only way JAX knows how to handle numerical data. I'm not sure how I'd approach changing this for cases where larger integers are required.

Neither of these is really a viable solution, and I can't think of any other alternatives.

**Comment by jakevdp** [comment]
The above shouldn't stop us from fixing the issue for smaller types, though: for example things like `random.randint(key, shape, minval=0, maxval=256, dtype='uint8')` should be workable with a few tweaks. But we would not be able to error on `maxval > 256` as numpy does, because JAX cannot branch on non-static values.

**Comment by jakevdp** [comment]
I started some work on this in #5868... it gets quite complicated to handle all the possible corner cases!

**Comment by hawkinsp** [comment]
Could we fix this by adding an `inclusive`/`exclusive` option on the API?

**Comment by jakevdp** [comment]
> Could we fix this by adding an inclusive/exclusive option on the API?

That's an option, although unless users know to think about it, things like `randint(key, shape, 0, 256, 'uint8')` will continue to silently return zeros.

**Comment by hawkinsp** [comment]
We should report (everywhere) when we see integer overflow casting Python integers to typed integers. Someone even started on that in the past.

**Comment by jakevdp** [comment]
That would be useful - I think currently we only get that via numpy when we overflow int64.

**Comment by jakevdp** [comment]
Also, even if we go the `inclusive / exclusive` route, I've come across some other corner cases that we'll need to fix as well, for example:
```python
from jax import random

random.randint(random.PRNGKey(0), (10,), 254, 254, dtype='uint8')
# Buffer([254, 254, 254, 254, 254, 254, 254, 254, 254, 254], dtype=uint8)

random.randint(random.PRNGKey(0), (10,), 255, 255, dtype='uint8')
# Buffer([111, 139,  85, 222,   6, 223, 108,  27,  20, 191], dtype=uint8)
```

**Comment by jakevdp** [comment]
Related: https://github.com/google/jax/issues/2006#issuecomment-575273556

**Comment by jakevdp** [comment]
It looks like the fix just made it into version 0.2.12. Thanks for reporting the error!
